File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Topic/FoxgloveUnsubscribe.swift
//
//  FoxgloveUnsubscribe.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// Unsubscribe メッセージ（クライアント → サーバー）
/// 購読の停止をリクエストする
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#unsubscribe
struct FoxgloveUnsubscribe: Codable {
    let op: FoxgloveMessageOperation
    let subscriptionIds: [UInt32]

    init(op: FoxgloveMessageOperation = .unsubscribe, subscriptionIds: [UInt32]) {
        self.op = op
        self.subscriptionIds = subscriptionIds
    }

    /// JSON 文字列にエンコードする
    func toJsonString() -> String? {
        guard let jsonData = try? JSONEncoder().encode(self),
              let jsonString = String(data: jsonData, encoding: .utf8) else {
            return nil
        }
        return jsonString
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/foxglove_bridge/topic/FoxgloveUnsubscribe.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.foxglove_bridge.topic

import com.shusuke.raspberry_pi_client.data.infrastructure.foxglove_bridge.common.FoxgloveMessageOperation
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Unsubscribe メッセージ（クライアント → サーバー）
 *
 * 購読の停止をリクエストする。
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#unsubscribe
 */
@Serializable
data class FoxgloveUnsubscribe(
    val op: FoxgloveMessageOperation = FoxgloveMessageOperation.Unsubscribe,
    @SerialName("subscriptionIds")
    val subscriptionIds: List<UInt> = emptyList(),
)
